paymentsheet/src/main/java/com/stripe/android/paymentsheet/paymentdatacollection/bacs/BacsMandateConfirmationViewModel.kt

```diff
@@ -0,0 +1,124 @@
+
+internal class BacsMandateConfirmationViewModel constructor(
+    private val args: Args
+) : ViewModel() {
+    private val _effect = MutableSharedFlow<BacsMandateConfirmationEffect>()
+    val effect: SharedFlow<BacsMandateConfirmationEffect>
+        get() = _effect
+
+    private val _viewState = MutableStateFlow(
+        BacsMandateConfirmationViewState(
+            email = args.email,
+            nameOnAccount = args.nameOnAccount,
+            sortCode = args.sortCode.chunked(size = 2).joinToString(separator = "-"),
+            accountNumber = args.accountNumber,
+            supportAddressAsHtml = buildAddressAsHtml(),
+            debitGuaranteeAsHtml = buildGuarantee()
+        )
+    )
+    val viewState: StateFlow<BacsMandateConfirmationViewState>
+        get() = _viewState
+
+    fun handleViewAction(action: BacsMandateConfirmationViewAction) {
+        when (action) {
+            is BacsMandateConfirmationViewAction.OnConfirmPressed -> onConfirmPress()
+            is BacsMandateConfirmationViewAction.OnCancelPressed -> onCancelPress()
+        }
+    }
+
+    private fun onConfirmPress() {
+        viewModelScope.launch {
+            _effect.emit(BacsMandateConfirmationEffect.CloseWithResult(BacsMandateConfirmationResult.Confirmed))
+        }
+    }
+
+    private fun onCancelPress() {
+        viewModelScope.launch {
+            _effect.emit(BacsMandateConfirmationEffect.CloseWithResult(BacsMandateConfirmationResult.Cancelled))
+        }
+    }
+
+    private fun buildAddressAsHtml(): String {
+        val defaultAddress = args.defaultAddress
+
+        return buildString {
+            append(defaultAddress.lineOne)
+            append("<br>")
+            append(defaultAddress.lineTwo)
+            append("<br>")
+            append("<a href=\"mailto:${defaultAddress.supportEmail}\">")
+            append(defaultAddress.supportEmail)
+            append("</a>")
+        }
+    }
+
+    private fun buildGuarantee(): String {
+        return buildString {
+            append("<a href=\"${args.guarantee.url}\">")
+            append(args.guarantee.name)
+            append("</a>")
+        }
+    }
+
+    data class Args(
+        val email: String,
+        val nameOnAccount: String,
+        val sortCode: String,
+        val accountNumber: String,
+        val guarantee: Guarantee,
+        val defaultAddress: DefaultAddress,
+        val defaultPayer: String
+    ) {
+        data class Guarantee(
+            val name: String,
+            val url: String
+        )
+
+        data class DefaultAddress(
+            val lineOne: String,
+            val lineTwo: String,
+            val supportEmail: String,
+        )
+    }
+
+    class Factory(
+        private val application: Application,
+        private val args: BacsMandateConfirmationContract.Args
+    ) : ViewModelProvider.Factory {
+        @Suppress("UNCHECKED_CAST")
+        override fun <T : ViewModel> create(modelClass: Class<T>): T {
```

Comment:
We should implement `create(Class, CreationExtras)` instead. `CreationExtras` give us access to the `Application`, so we don’t have to provide it ourselves.